sys/drv/rtl8139.c

```diff
@@ -21,7 +22,29 @@ static int rtl8139_probe(device_t *dev) {
   return pci_device_match(pcid, RTL8139_VENDOR_ID, RTL8139_DEVICE_ID);
 }
 
+static int rtl_reset(rtl8139_state_t *state) {
+  bus_write_1(state->regs, RL_8139_CFG1, RL_CFG1_PWRON);
+  bus_write_1(state->regs, RL_COMMAND, RL_CMD_RESET);
+  for (int i = 0; i < RL_TIMEOUT; i++) {
+    if (!(bus_read_1(state->regs, RL_COMMAND) & RL_CMD_RESET))
+      return 0;
+  }
+
+  return 1;
+}
+
 static int rtl8139_attach(device_t *dev) {
+  assert(dev->parent->bus == DEV_BUS_PCI);
+  rtl8139_state_t *state = dev->state;
+
+  state->regs = bus_alloc_resource_any(dev, RT_MEMORY, 1, RF_ACTIVE);
+
+  if (state->regs == NULL)
+    return -1;
+
+  if (rtl_reset(state))
```

Comment:
It's interesting. I presume that reset initializes some internal state (like counters and registers). If it fails then what guarantee do you have about consistency of the state?
It would be nice to add a comment which explains that this kind of failure isn't fatal.